zerobuf/BaseVector.h

```diff
@@ -29,15 +31,53 @@ class BaseVector
      * @param alloc The parent allocator that contains the data.
      * @param index Index of the vector in the parent allocator dynamic storage
      */
-    BaseVector( A* alloc, size_t index );
+    BaseVector( A* alloc,
+                size_t index,
+                size_t staticSize = sizeof( T ));
     virtual ~BaseVector() {}
 
-    const T& operator[] ( size_t index ) const;
+    template<class Q = T>
+    typename std::enable_if<!std::is_base_of<Zerobuf,Q>::value, Q>::type
+    operator[] ( const size_t index ) const
+    {
+        if( index >= size( ))
+            throw std::runtime_error( "Vector out of bounds read" );
+
+        return data()[ index ];
+    }
+
+    template<class Q = T>
+    typename std::enable_if<std::is_base_of<Zerobuf,Q>::value, Q>::type
+    operator[] ( const size_t index ) const
+    {
+        if( index >= size( ))
+            throw std::runtime_error( "Vector out of bounds read" );
+
+        const size_t dynOff = (size_t)_parent->template getDynamicPtr<uint8_t>( _index )
+                              - (size_t)_parent->getData();
+
+        ConstNonMovingSubAllocator* constRef =
+                new ConstNonMovingSubAllocator( static_cast<const NonMovingBaseAllocator*>( _parent ),
+                                                dynOff + index * _elemSize,
+                                                0,
+                                                _elemSize );
+
+        const Q constRefObj( constRef );
```

Comment:
I did not like the code below ( needs implicit knowledge of there is a copyable buffer ). Lets discuss what will be the nicer way of having it. A way to do that is creating a new NonMovableBuffer and copying the prev buffer through a copy constructor.